File: src/main/java/com/exampledigisphere/vitapetcare/config/security/UserDetailsImpl.java
package com.exampledigisphere.vitapetcare.config.security;

import com.exampledigisphere.vitapetcare.admin.user.domain.User;
import com.exampledigisphere.vitapetcare.auth.roles.RoleAuthorityMapper;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

import java.util.Collection;
import java.util.stream.Collectors;

public class UserDetailsImpl implements UserDetails {
  private final User user;

  public UserDetailsImpl(final User user) {
    this.user = user;
  }

  @Override
  public Collection<? extends GrantedAuthority> getAuthorities() {
    return RoleAuthorityMapper.getAuthorities(this.user.getRole())
      .stream()
      .map(SimpleGrantedAuthority::new)
      .collect(Collectors.toSet());
  }

  @Override
  public String getPassword() {
    return user.getPassword();
  }

  @Override
  public String getUsername() {
    return user.getEmail();
  }

  @Override
  public boolean isAccountNonExpired() {
    return UserDetails.super.isAccountNonExpired();
  }

  @Override
  public boolean isAccountNonLocked() {
    return UserDetails.super.isAccountNonLocked();
  }

  @Override
  public boolean isCredentialsNonExpired() {
    return UserDetails.super.isCredentialsNonExpired();
  }

  @Override
  public boolean isEnabled() {
    return UserDetails.super.isEnabled();
  }
}

File: src/main/kotlin/com/exampledigisphere/vitapetcare/config/security/UserDetailsImpl.kt
package com.exampledigisphere.vitapetcare.config.security

import com.exampledigisphere.vitapetcare.admin.user.domain.User
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails

class UserDetailsImpl(private val user: User?) : UserDetails {

  override fun getAuthorities(): MutableCollection<out GrantedAuthority> {
    val authorities = mutableListOf<GrantedAuthority>()
    authorities.add(SimpleGrantedAuthority("ROLE_${user?.role?.name}"))
    user?.role?.permissions?.forEach { perm ->
      authorities.add(SimpleGrantedAuthority(perm))
    }
    return authorities
  }

  override fun getPassword(): String = user?.password!!
  override fun getUsername(): String = user?.email!!

  override fun isAccountNonExpired(): Boolean = true
  override fun isAccountNonLocked(): Boolean = true
  override fun isCredentialsNonExpired(): Boolean = true
  override fun isEnabled(): Boolean = true
}
